Update logging with new abstraction/configuration (#7271)
umbraco/UmbracoDocs: 16/umbraco-cms/fundamentals/backoffice/logviewer.md

If you frequently use a custom query, you can save it for quick access. Type your query in the search box and click the heart icon to save it with a friendly name. Saved queries are stored in the `umbracoLogViewerQuery` table in the database.

## Implementing Your Own Log Viewer Source

> Umbraco allows you to implement a customn `ILogViewerRepository` to fetch logs from alternative sources, such as **Azure Table Storage**.

Reviewer: Isn't it necessary to also implement a custom `ILogViewerService` too, due to the `CanViewLogs` method in `LogViewerService`?

        if (properties is not null)
        {
            foreach (KeyValuePair<string, LogEventPropertyValue> property in properties)
            {
                string? value;

                if (property.Value is ScalarValue scalarValue)
                {
                    value = scalarValue.Value?.ToString();
                }
                else if (property.Value is StructureValue structureValue)
                {
                    var textWriter = new StringWriter();
                    structureValue.Render(textWriter);
                    value = textWriter.ToString();
                }
                else
                {
                    value = _jsonSerializer.Serialize(property.Value);
                }

                result.Add(property.Key, value);
            }
        }

        return result;
    }

    public class AzureTableLogEntity : ITableEntity
    {
        public required string Data { get; set; }

        public required string PartitionKey { get; set; }

        public required string RowKey { get; set; }

        public DateTimeOffset? Timestamp { get; set; }

        public ETag ETag { get; set; }
    }
}
```

> Azure Table Storage requires entities to implement the `ITableEntity` interface. Since Umbraco’s default log entity does not implement this, a custom entity (`AzureTableLogEntity`) must be created to ensure logs are correctly fetched.

Reviewer: ```suggestion
Azure Table Storage requires entities to implement the `ITableEntity` interface. Since Umbraco's default log entity does not implement this, a custom entity (`AzureTableLogEntity`) must be created to ensure logs are correctly fetched.
```

### Creating a custom log viewer service

> The next thing we need to do is create a new service that amongs other things is responsible to figure out whether a certain log query is allowed.

Reviewer: ```suggestion
The next thing we need to do is create a new service that amongst other things is responsible for figuring out whether a provided log query is allowed.
```

```csharp
public class AzureTableLogsService : LogViewerServiceBase
{

>     private readonly ILogViewerRepository _logViewerRepository;

Reviewer: Maybe we could expose this on the base class, so you don't need to save it into a private field here too?

        ICoreScopeProvider provider,
        ILogViewerRepository logViewerRepository)
        : base(logViewerQueryRepository, provider, logViewerRepository)
    {
        _logViewerRepository = logViewerRepository;
    }

    // Change this to what you think is sensible
    // as an example we check whether more than 5 days off logs are requested
    public override Task<Attempt<bool, LogViewerOperationStatus>> CanViewLogsAsync(LogTimePeriod logTimePeriod)
    {
        return logTimePeriod.EndTime - logTimePeriod.StartTime < TimeSpan.FromDays(5)
            ? Task.FromResult(Attempt.SucceedWithStatus(LogViewerOperationStatus.Success, true))
            : Task.FromResult(Attempt.FailWithStatus(LogViewerOperationStatus.CancelledByLogsSizeValidation, false));
    }

    public override ReadOnlyDictionary<string, LogLevel> GetLogLevelsFromSinks()
    {
        var configuredLogLevels = new Dictionary<string, LogLevel>
        {
            { "Global", GetGlobalMinLogLevel() },
            { "AzureTableStorage", _logViewerRepository.RestrictedToMinimumLevel() },
        };

        return configuredLogLevels.AsReadOnly();
    }
}
```

### Register implementation

> Umbraco needs to be made aware that there is a new implementation of an `ILogViewerRepository` to register. We also need to replace the default JSON LogViewer that is shipped in the core of Umbraco.

Reviewer: ```suggestion
Umbraco needs to be made aware that there is a new implementation of an `ILogViewerRepository` and an `ILogViewerService` to register. These need to replace the default ones that are shipped with Umbraco.
```

```csharp
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Infrastructure.DependencyInjection;
using Umbraco.Cms.Core.Services;

namespace My.Website;

public class AzureTableLogsComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
        {
            builder.Services.AddUnique<ILogViewerRepository, AzureTableLogsRepository>();
            builder.Services.AddUnique<ILogViewerService, AzureTableLogsService>();
        }
    }
}
```

### Configuring Logging to Azure Table Storage

> With the above two classes, the setup is in place to view logs from an Azure Table. However, logs are not yet persisted into the Azure Table Storage account. To enable persistence, configure the Serilog logging pipeline to store logs in Azure Table Storage.

Reviewer: "two classes" --> "three classes"

Reviewer: Applied.

using Azure.Data.Tables;
using Serilog.Events;
using Serilog.Formatting.Compact.Reader;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.Logging.Viewer;
using Umbraco.Cms.Core.Serialization;
using Umbraco.Cms.Core.Services;
using Umbraco.Cms.Infrastructure.Logging.Serilog;
using Umbraco.Cms.Infrastructure.Services.Implement;

namespace My.Website;

public class AzureTableLogsRepository : LogViewerRepositoryBase
{
    private readonly IJsonSerializer _jsonSerializer;

    public AzureTableLogsRepository(UmbracoFileConfiguration umbracoFileConfig, IJsonSerializer jsonSerializer) : base(
        umbracoFileConfig)
    {
        _jsonSerializer = jsonSerializer;
    }

    protected override IEnumerable<ILogEntry> GetLogs(LogTimePeriod logTimePeriod, ILogFilter logFilter)
    {

>         // This example uses a connetionstring compatible with the Azurite emulator

Reviewer: ```suggestion
        // This example uses a connection string compatible with the Azurite emulator
```
